src/probe_modules/module_dns.c

```diff
@@ -0,0 +1,1043 @@
+#include "../../lib/xalloc.h"
+#include "probe_modules.h"
+#include "packet.h"
+#include "logger.h"
+#include "module_dns.h"
+#include "module_udp.h"
+#include "../fieldset.h"
+
+#define DNS_SEND_LEN 512    // This is arbitrary 
+#define UDP_HEADER_LEN 8
+#define PCAP_SNAPLEN 3000   // This is even more arbitrary
+#define UNUSED __attribute__((unused))
+#define MAX_QTYPE 255
+#define ICMP_UNREACH_HEADER_SIZE 8
+#define BAD_QTYPE_STR "BAD QTYPE"
+#define BAD_QTYPE_VAL -1
+#define MAX_LABEL_RECURSION 10
+
+// Note: each label has a max length of 63 bytes. So someone has to be doing
+// something really annoying. Will raise a warning.
+// THIS INCLUDES THE NULL BYTE
+#define MAX_NAME_LENGTH 512 
+
+typedef uint8_t bool;
+
+// zmap boilerplate
+probe_module_t module_dns;
+static int num_ports;
+
+const char default_domain[] = "www.google.com";
+const uint16_t default_qtype = DNS_QTYPE_A; 
+
+static char *dns_packet = NULL;
+static uint16_t dns_packet_len = 0; // Not including udp header
+static uint16_t qname_len = 0;
+static char* qname = NULL;
+static uint16_t qtype = 0;
+
+/* Array of qtypes we support. Jumping through some hoops (1 level of 
+ * indirection) so the per-packet processing time is fast. Keep this in sync with:
+ * dns_qtype (.h)
+ * qtype_strid_to_qtype (below)
+ * qtype_qtype_to_strid (below, and _setup_qtype_str_map())
+ */
+const char *qtype_strs[] = {
+        "A",
+        "NS",
+        "CNAME",
+        "SOA",
+        "PTR",
+        "MX",
+        "TXT",
+        "AAAA",
+        "RRSIG",
+        "ALL"
+};
+
+const dns_qtype qtype_strid_to_qtype[] = { DNS_QTYPE_A, DNS_QTYPE_NS, DNS_QTYPE_CNAME, 
+        DNS_QTYPE_SOA, DNS_QTYPE_PTR, DNS_QTYPE_MX, DNS_QTYPE_TXT, 
+        DNS_QTYPE_AAAA, DNS_QTYPE_RRSIG, DNS_QTYPE_ALL
+};
+
+#define DNS_QR_ANSWER   1
+
+int8_t qtype_qtype_to_strid[256] = { BAD_QTYPE_VAL };
+
+void _setup_qtype_str_map() 
+{ 
+    qtype_qtype_to_strid[DNS_QTYPE_A] = 0;
+    qtype_qtype_to_strid[DNS_QTYPE_NS] = 1;
+    qtype_qtype_to_strid[DNS_QTYPE_CNAME] = 2;
+    qtype_qtype_to_strid[DNS_QTYPE_SOA] = 3;
+    qtype_qtype_to_strid[DNS_QTYPE_PTR] = 4;
+    qtype_qtype_to_strid[DNS_QTYPE_MX] = 5;
+    qtype_qtype_to_strid[DNS_QTYPE_TXT] = 6;
+    qtype_qtype_to_strid[DNS_QTYPE_AAAA] = 7;
+    qtype_qtype_to_strid[DNS_QTYPE_RRSIG] = 8;
+    qtype_qtype_to_strid[DNS_QTYPE_ALL] = 9;
+}
+
+uint16_t _qtype_str_to_code(char* str) 
+{
+    for (int i = 0; i < (int) (sizeof(qtype_strs)/sizeof(qtype_strs[0])); i++) {
+        if (strcmp(qtype_strs[i], str) == 0)
+            return qtype_strid_to_qtype[i];
+    }
+    return 0;
+}
+
+
+static uint16_t _domain_to_qname(char** qname_handle, char* domain) 
+{   
+    // String + 1byte header + null byte
+    uint16_t len = strlen(domain) + 1 + 1;
+    char* qname = xmalloc(len);
+    // Add a . before the domain. This will make the following simpler.
+    qname[0] = '.';
+    // Move the domain into the qname buffer.
+    memcpy(qname + 1, domain, strlen(domain));
```

Comment:
Do you expect a nullbyte? You are not guaranteed this with a memcpy. Why not use strcpy?